src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc

```diff
@@ -170,547 +161,472 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
+  if (!_sdf->HasElement("force_length"))
+  {
+    ignlog << "Missing parameter <force_length>, "
+      << "setting to " << this->dataPtr->forceLength << std::endl;
+  }
+  else
+  {
+    if (_sdf->Get<double>("force_length") < 0)
     {
-      if (_ecm.EntityHasComponentType(sensor,
-                                      components::DepthCamera::typeId))
-      {
-        depthCameraInLink = true;
-        depthCameraSdf =
-          _ecm.Component<components::DepthCamera>(sensor)->Data().Element();
-      }
-
-      if (_ecm.EntityHasComponentType(sensor,
-                                      components::ContactSensor::typeId))
-      {
-        contactSensorInLink = true;
-      }
+      ignwarn << "Parameter <force_length> must be positive, "
+        << "setting to " << this->dataPtr->forceLength << std::endl;
     }
-
-    if (!depthCameraInLink || !contactSensorInLink ||
-      sensorsInsideLink.size() != 2)
+    else
     {
-      ignerr << "Link must have 1 Depth Camera sensor and "
-        << "1 Contact sensor" << std::endl;
-      return;
+      this->dataPtr->forceLength = _sdf->Get<double>("force_length");
     }
+  }
+
+  // Get the size of the sensor from the SDF
+  // If there's no <collision> specified inside <link>, a default one
+  // is set
+  this->dataPtr->sensorSize =
+    _sdf->GetParent()->GetElement("link")->GetElement("collision")->
+    GetElement("geometry")->GetElement("box")->
+    Get<ignition::math::Vector3d>("size");
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePlugin::PreUpdate(const UpdateInfo &_info,
+  EntityComponentManager &_ecm)
+{
+  IGN_PROFILE("TouchPluginPrivate::PreUpdate");
+
+  // Nothing left to do if paused
+  if (_info.paused)
+    return;
 
-    // Store depth camera update rate and offset from model
-    auto offset = depthCameraSdf->Get<ignition::math::Pose3d>("pose");
-    this->cameraUpdateRate = 1;
-    if (depthCameraSdf->HasElement("update_rate"))
+  if (!this->dataPtr->initialized)
+  {
+    // We call Load here instead of Configure because we can't be guaranteed
+    // that all entities have been created when Configure is called
+    this->dataPtr->Load(_ecm);
+    this->dataPtr->initialized = true;
+  }
+
+  // This is not an "else" because "initialized" can be set
+  // in the if block above
+  if (this->dataPtr->initialized)
+  {
+    // Get the first object being touched by the sensor
+    // We assume there's only one object being touched
+    auto csd = _ecm.Component<components::ContactSensorData>(
+      this->dataPtr->sensorCollisionEntity);
+    for (const auto &contact : csd->Data().contact())
     {
-      this->cameraUpdateRate = depthCameraSdf->Get<float>("update_rate");
+      this->dataPtr->objectCollisionEntity =
+        contact.collision2().id();
+      break;
     }
 
+    // Get the Depth Camera pose
+    ignition::math::Pose3d sensorPose =
+      _ecm.Component<components::Pose>(
+        this->dataPtr->model.Entity())->Data();
```

Comment:
> The comment says depth camera pose, but this is actually getting the model pose

Yes, I was actually referring to the model pose.

> I think the frame name is just the model's name, but to be sure, you'll want to test this

I thought the same, but I've actually tested removing the attribute, setting it to `ground_plane` and even names not related to simulation and it seems the plugin keeps doing OK... weird. I think that an `ignwarn` if the model's name and `relative_to` value are not equal is still a good idea.

Reply:
Huh! That is weird. I'm seeing sdformat tests that in `ign_TEST.cc` https://github.com/osrf/sdformat/blob/master/src/ign_TEST.cc#L499

Looks like it wants link, joint, or frame name. So probably use the link name instead of model name for the `ignwarn` check.

Reply:
The `relative_to` attribute of poses for sensors is resolved during [creation]( https://github.com/ignitionrobotics/ign-gazebo/blob/fd37a777122e214b8051ea404eaaa151d110b52e/src/SdfEntityCreator.cc#L519) and it is expressed relative to the parent element (link or joint) that contains the sensor, so there is no need to handle the `relative_to` attribute here. This is true for all other SDFormat elements as well so long as their pose is defined in the SDFormat spec. i.e, this doesn't include `<pose>` elements inside `<plugin>` for example.

Reply:
Cool, thanks for the clarification!

Reply:
e216bb2